xarray/backends/api.py

```diff
@@ -147,6 +149,123 @@ def _get_default_engine(path: str, allow_remote: bool = False) -> T_NetcdfEngine
+                if isinstance(concat_dim, DataArray):
+                    warnings.warn(
+                        "In a future version of xarray default value for data_vars  will change from "
+                        "data_vars='all' to data_vars='minimal'. This is likely to lead to "
+                        "different results when using a `DataArray` as the concat_dim. "
+                        "To opt in to new defaults and get rid of these warnings now "
+                        "use `set_option(use_new_open_mfdataset_kwargs=True) or set data_vars explicitly.",
+                        category=FutureWarning,
+                        stacklevel=2,
+                    )
+                    break
+                # dimensions without coordinate arrays are fine
+                if concat_dim is None or concat_dim not in ds:
+                    continue
+                concat_dim_min = ds[concat_dim].min()
+                concat_dim_max = ds[concat_dim].max()
+                for i, (min, max) in enumerate(concat_dim_ranges.get(concat_dim, [])):
+                    if concat_dim_min <= max and concat_dim_max >= min:
+                        overlaps_at_index.append(i)
+                concat_dim_ranges[concat_dim] = [
+                    *concat_dim_ranges.get(concat_dim, []),
+                    (concat_dim_min, concat_dim_max),
+                ]
+            # check if this dataset overlaps with one other dataset on all concat_dims
+            if (
+                len(overlaps_at_index) == len(concat_dims)
+                and len(set(overlaps_at_index)) == 1
+            ):
+                warnings.warn(
+                    "In a future version of xarray default value for data_vars  will change from "
+                    "data_vars='all' to data_vars='minimal'. This is likely to lead to "
+                    "different results when multiple datasets have overlapping concat_dim values. "
+                    "To opt in to new defaults and get rid of these warnings now "
+                    "use `set_option(use_new_open_mfdataset_kwargs=True) or set data_vars explicitly.",
+                    category=FutureWarning,
+                    stacklevel=2,
+                )
+                break
+
+    if "compat" not in input_kwargs and input_kwargs.get("data_vars") == "different":
+        # make sure people set compat explicitly for this case
+        warnings.warn(
+            "In a future version of xarray default value for compat  will change from "
+            "compat='no_conflicts' to compat='override'. This is likely to lead to "
+            "failures when used with data_vars='different'. The recommendation is to "
+            "set compat='no_conflicts' explicitly for this case.",
+            category=FutureWarning,
+            stacklevel=2,
+        )
+
+    if "join" not in input_kwargs:
+        # TODO: Figure out how conditions that xarray/tests/test_backends.py::TestDask::test_encoding_mfdataset
```

Comment:
I am having a hard time figuring out how to catch the conditions that make xarray/tests/test_backends.py::TestDask::test_encoding_mfdataset fail with the new join kwarg. It seems like it has to do with the concat_dim in each dataset having a different encoding, but the values of the concat_dim in each dataset doe not overlap. So I'm a little wondering if the align index function is too strict.